taf/updater/updater_pipeline.py

```diff
@@ -796,90 +819,95 @@ def determine_start_commits(self):
-                self.state.targets_data_by_auth_commits = (
-                    self.state.users_auth_repo.targets_data_by_auth_commits(
-                        self.state.auth_commits_since_last_validated
-                    )
-                )
-                # start validation from the beginning, so also removed
-                # information about the top commits of user's repositories
-                for repository in self.state.users_target_repositories.values():
-                    for branch in self.state.old_heads_per_target_repos_branches[
-                        repository.name
-                    ]:
-                        self.state.old_heads_per_target_repos_branches[repository.name][
-                            branch
-                        ] = None
+                self._update_state_for_initial_sync()
+                self._reset_target_repositories()
 
             return UpdateStatus.SUCCESS
         except Exception as e:
             self.state.errors.append(e)
             self.state.event = Event.FAILED
             return UpdateStatus.FAILED
 
+    def _is_repository_in_sync(self, repository, last_validated_commit_data):
+        current_branch = last_validated_commit_data.get(
+            "branch", repository.default_branch
+        )
+        last_validated_commit = last_validated_commit_data["commit"]
+
+        if repository.is_bare_repository:
+            try:
+                top_commit_of_branch = repository.top_commit_of_remote_branch(
+                    current_branch
+                )
+            except GitError:
+                # Check if the user's local repository has the branch
+                if current_branch in repository.branches:
+                    repository.create_branch(current_branch)
+                    # Create the branch in the temp repository
+                    top_commit_of_branch = repository.top_commit_of_branch(
+                        current_branch
+                    )
+                else:
+                    return False
+        else:
+            if not repository.branch_exists(current_branch, include_remotes=False):
+                return False
+            top_commit_of_branch = repository.top_commit_of_branch(current_branch)
+
+        if top_commit_of_branch != last_validated_commit:
+            # check if top commit is newer (which is fine, it will be validated)
+            # or older, meaning that the authentication repository contains
+            # additional commits, so it would be necessary to find older auth repo
+            # commit and start the validation from there
+            if current_branch not in repository.branches_containing_commit(
+                last_validated_commit
+            ):
+                return False
+            self.state.old_heads_per_target_repos_branches[repository.name][
+                current_branch
+            ] = last_validated_commit
+
+        return True
+
+    def _reset_target_repositories(self):
```

Comment:
I don't think that we need to call `clean_and_reset()` again. The idea behind this code was to update the state, without making any changes to the repositories. Basically, to make sure that we'll validate everything from the beginning. `clean_and_reset` should get called only if we detect that there a repository contains uncommitted/unpushed changes. You are right though, with this `--force` flag, we will need to make some additional changes if repositories are not in sync. I'll update the issue with the next steps